src/cudamatrix/cu-kernels.cu

```diff
@@ -2630,6 +2630,467 @@ static void _diff_log_softmax(const MatrixDim in_deriv_dim,
 }
 
 
+/**
+ this function computes the core part of the LSTM nonlinearity.
+ @param [in] in      A matrix, of dimension num_rows by 5*cell_dim
+                     (i.e. its num-cols must be a multiple of 5).
+                     The column-space is interpreted as 5
+                     consecutive blocks, each of dimension cell_dim,
+                     which we name:
+                     (i_part, f_part, c_part, o_part, c_{t-1}).
+ @param [in] params  A matrix, of dimension 3 by cell_dim,
+                     with rows containing the 3 diagonal parameter matrices
+                     used in LSTMs, namely
+                     w_{ic}, w_{fc} and w_{oc}.
+ @param [out] out    A matrix, of dimension num_rows by 2*cell_dim.
+                     The quantities c_t and m_t respectively are put there
+                     (in two blocks of column-dimension cell_dim),
+                     according to the following equations:
+
+                     i_t = Sigmoid(i_part + w_{ic}*c_{t-1})
+                     f_t = Sigmoid(f_part + w_{fc}*c_{t-1})
+                     c_t = f_t*c_{t-1} + i_t * Tanh(c_part)
+                     o_t = Sigmoid(o_part + w_{oc}*c_t)
+                     m_t = o_t * Tanh(c_t)
+
+We use 1D thread block with CU1DBLOCK threads.
+It works best when cell_dim is a multiple of CU1DBLOCK.
+We use 1d Grid. Each block is working on one row of the in and out matrices.
+*/
+template<typename Real>
```

Comment:
I think template is required so that we can run both `double` and `float` tests at the same time.